Add button to fix issue 60

david-mears-dfe opened the pull request:
Fixes https://github.com/microcovid/microcovid/issues/60

## Description

(Please note that I am a Ruby developer, not a React developer, so there may be whoopsies.)

This displays a `or enter data` button if no top location has been set. The button selects `Select a location or enter data...` on the top location drop-down. The button does not exist when `Select a location or enter data...` is selected.

'Step 1 - Choose a location' has been edited to say 'Step 1 - Location Risk' (to match the Person Risk and Activity Risk sections, and to avoid implying that the only method of getting the location risk is by choosing a location from the drop-down.)

## Screenshots

<img width="397" alt="Screenshot 2020-10-09 at 14 53 27" src="https://user-images.githubusercontent.com/60350599/95591722-98efc580-0a3f-11eb-9f42-37eb06419f08.png">
<img width="397" alt="Screenshot 2020-10-09 at 14 53 36" src="https://user-images.githubusercontent.com/60350599/95591723-99885c00-0a3f-11eb-8b3e-60cbba5b5289.png">

david-mears-dfe commented:
@beshaya @dominic have you seen this?

beshaya commented:
The team is working on a major overhaul of the calculator UX, and we are including something similar. Check out the ux-redesign branch or the preview at https://ux-redesign--microcov.netlify.app/ for what's happening.

david-mears-dfe commented:
> The team is working on a major overhaul of the calculator UX, and we are including something similar. Check out the ux-redesign branch or the preview at https://ux-redesign--microcov.netlify.app/ for what's happening.

Thanks for the reply @beshaya. Would you welcome a similar PR to that branch?

david-mears-dfe commented:
(As I don't see an equivalent button on that preview url)

beshaya commented:
Yeah, you're right. I think the button is easier to see. @blanchardjeremy what do you think?

blanchardjeremy commented:
Hmmm. I agree that button is easier to see. This is a convo we've been having over the past few weeks. 

I opted for a drop-down menu item because I don't think that many users actually want to enter data manually. It's a fairly advanced operation. I imagine people would opt for country level data even if we don't have enough their particular region.

But I have no data to back up my guess! 😂 

@david-mears-dfe are you someone who's entering data manually? If so can you tell us more about your use case? And why you'd prefer a button over the drop-down menu item?

blanchardjeremy commented:
Oh I guess I should also ask: did you see that this feature does exist as a drop-down item? Maybe that wasn't already clear.

david-mears-dfe commented:
> Oh I guess I should also ask: did you see that this feature does exist as a drop-down item? Maybe that wasn't already clear.

It wasn't. There is a lot of detail about this described in the issue this PR fixes. https://github.com/microcovid/microcovid/issues/60.

> are you someone who's entering data manually? If so can you tell us more about your use case? And why you'd prefer a button over the drop-down menu item?

I'm entering data manually because averaging across England is not accurate enough. This variance makes a large difference to what activities are tolerable. From the UK government's point of view, there are 3 very different 'tiers' of lockdown across the country according to local prevalence. In Scotland they use 5 tiers.

I've now 'advanced' to using the spreadsheet, but in my initial forays into microcovid (a necessary precursor to using the spreadsheet?) I used the calculator.

blanchardjeremy commented:
@david-mears-dfe is this still something you're interested in submitting a PR for? If so, let us know.

david-mears-dfe commented:
Yes it is, I'm going to look at it today.

david-mears-dfe commented:
I can't see anything called MANUAL_DATA_KEY in the code, but my plan is to have the button simply select the drop-down option (as in this original PR).

david-mears-dfe commented:
Ok I've gone ahead and updated this PR.

Each change is divided into very atomic commits so that I can easily discard any changes you don't like.

Then I can squash the commits if you would like.

blanchardjeremy commented:
@david-mears-dfe @david-mears - looking good so far. 

Is the main use case you want to solve "how to switch back to manual location data"? I ask because I figured you would also want the button to be visible when nothing was selected in either drop-down. So that folks could find it more easily. 

Also I suggest we move the button below the second drop-down? I think that fits better visually and contextually. Let me know. 

Thanks for doing this!

david-mears-dfe commented:
Re-updated. I have updated the behaviour as you suggest and the position. I've also added a button to return to inputting a location which is visible in 'enter data' mode.

david-mears-dfe commented:
@blanchardjeremy Would you have time to take a look this week? I don't want this to get too far out of step with the latest master.

blanchardjeremy commented:
Thanks for following up. At this point, I'm not sure I'm the best person to see this one through.  I find the current layout confusing, but I don't have a suggestion for how to make it better.  I don't want to be a blocker on forward progress. So I'll ask if anyone else on the team might be willing to chime in. Maybe @beshaya

blanchardjeremy commented:
I propose we close this as there hasn't been any movement on it in a while.

david-mears-dfe commented:
I'm sending some feedback on the contribution process to @catherio .

blanchardjeremy commented:
@david-mears-dfe Hey. Thanks for sending along feedback in that way. If you feel comfortable, I'd invite your feedback here as well so we can attempt to clear it up directly. I'll say upfront: I'm sorry that this hasn't been brought to completion.  I can understand it being frustrating to have something you want to contribute and it not being seen through. Perhaps it was short-sighted of me to propose closing this rather than giving it a clear path forward. I'd be open to hearing what you want to share and I'm sure the rest of the team would appreciate it too. Let us know what you feel comfortable with. (This PR seems as a good a place as any for this conversation, IMO. Or I suppose we could move it over to [Discussions](https://github.com/microcovid/microcovid/discussions))

david-mears-dfe commented:
Thanks @blanchardjeremy. I sent the feedback to Catherine as a precaution because I couldn't be sure you'd like to have the feedback public. Here's what I wrote.

> This is some hopefully constructive criticism for the microcovids team coming from a place of frustration.

> I think it might be worth the microcovids team thinking about how to make it easier to contribute changes to the project (or it might not, e.g. if the pandemic is over soon).

> After having it confirmed that a PR would be welcomed on this issue, I submitted the PR, and updated it in response to feedback, believing I was making a noticeable usability improvement to your site. But then I had my work go to waste because of a lack of timely engagement. If it hadn't gone to waste, I can confidently say I would have continued to make and submit fixes based on your issue backlog (instead, I've chosen other EA projects to contribute to).

> In the end, I wasted a lot of time. This was frustrating for me but not necessarily any one person's fault. From your perspective, I can understand that it's easily possible that the overhead of dealing with my PRs would have outweighed their benefit to users. And it's easy to overestimate how much capacity a team will have for dealing with external PRs, and to overlook designing a good process for contributors.

> One idea for what I would have preferred is: your team (or an initiative-taker) decided on a particular person to 'own' the PR, who then is empowered to decide at any point that it's not worth continuing with it, and who is responsible for communicating that to the author, and for avoiding the failure mode of the work being left there until it's incompatible or stale.

> Going forward, I think you could do something like that, rather than having a de facto halfway house policy that wastes time on both sides and goodwill on the contributors' side; or maybe you don't need a policy per se but would just benefit from being more aware of this blindspot/failure mode.

I'm open to re-updating the branch once more, iff you want me to and iff it's unlikely to go stale again.

blanchardjeremy commented:
@david-mears-dfe I just had a chance to read your message. Thank you for taking the time to share this. I really appreciate the thoughtful and compassionate way you shared it! Makes it easy to engage with and reply to. 👍

Since I've been the main person tracking this and engaging with you, I think most of how this went rests with me. I hear you when you say it's frustrating to have a PR invited and then dropped the way this was dropped. Sorry for my part in that, in "punting" it to the team without making sure someone picked up the ball. I think your suggestion about having a clear point person for each PR makes a lot of sense. Seems like a way to support people to have a successful experience contributing. It's important to me (and the team) that folks are supported to make the contributions that make sense to make, and that people not waste their time on things that aren't likely to be accepted.

blanchardjeremy commented:
As far as this PR goes, if you'd like to still contribute this, I"m down to be the point person for it!  I think the desired behavior that would resolve my earlier confusion would basically be something like:

* A button below both of the drop-downs that says "Switch to manual data entry mode" 
* Remove "enter data manually" from the drop-down menu. --- I could be swayed that it makes sense to have it in both places. But my first thought is that it seems confusing to me. Better to have it in just one place, IMO. The trick with implementing this approach will be that that field still needs to contain the MANUAL_DATA key so that it works with shareable URLs. So you'd have to add a bit of code in a few places to set that key into the field (and URL) and pull it out of the URL on page load.
* When you click the bottom, both dropdowns are hidden and button text changes to "Switch to pre-populated location data" (or something like that)

Those are my ideas. I'm open to revisions. Let me know your thoughts!

david-mears-dfe commented:
Thanks @blanchardjeremy. I'll take a look at the weekend.

Reminded of:
![image](https://user-images.githubusercontent.com/60350599/103831327-bd296580-5073-11eb-982e-fb2526f54094.png)

david-mears-dfe commented:
@blanchardjeremy I've added the behaviour you describe. It was less complicated than expected to preserve the shareable url functionality.

blanchardjeremy commented:
This is looking really good to me!
Only thing that seems off is this: when you're on automatic location data mode, it no longer shows the extra info (like cases last week). If you compre the deploy preview to the current live site, you should see the difference.

blanchardjeremy commented on src/components/calculator/PrevalenceControls.tsx:
```diff
@@ -249,78 +257,98 @@ export const PrevalenceControls: React.FunctionComponent<{
-          <div>{t('calculator.prevalence.cases_stable_or_decreasing')}</div>
-        ) : (
-          <PrevalenceField
-            id="precent-increase"
-            label={t('calculator.prevalence.percent_increase_in_cases')}
-            value={data.casesIncreasingPercentage}
-            unit="%"
-            setter={(value) => {
-              setter({ ...data, casesIncreasingPercentage: Number(value) })
-            }}
-            inputType="number"
-            min={0}
-            isEditable={!locationSet}
-            className="hide-number-buttons"
-          />
-        )}
-        {data.positiveCasePercentage === null ? (
-          <PrevalenceField
-            id="positive-test-rate"
-            label={t('calculator.prevalence.positive_case_percentage')}
-            value="no data available"
-            unit="%"
-            setter={(_value) => null}
-            inputType="text"
-            isEditable={false}
-          />
+        {!isManualEntryCurrently ? (
+          <PrevalenceResult data={data} />
         ) : (
-          <PrevalenceField
-            id="positive-test-rate"
```
I'd still like to see these fields show up even if automatic data location is being used:

![image](https://user-images.githubusercontent.com/32290/104152393-41415d00-5394-11eb-8318-89ed800922ca.png)

david-mears-dfe commented:
I fixed the issue you mention @blanchardjeremy .

While working on this I also noticed the following issue, both on this branch and on the live site:

When the location is set to 'Select location', and you click the Details drop down, the fields are editable. Is this intentional? I think it isn't, because if you fill the fields, it doesn't behave as if you are it manual data-entry mode. I think the fix will be to use `isManualDataCurrently` rather than `!locationSet` for the PrevalenceFields' isEditable attribute. I can submit this change in a separate PR as it's not related to this PR.

blanchardjeremy commented:
> While working on this I also noticed the following issue, both on this branch and on the live site:
> 
> When the location is set to 'Select location', and you click the Details drop down, the fields are editable. Is this intentional? I think it isn't, but I'm not sure. I think the fix will be to use `isManualDataCurrently` rather than `!locationSet` for the PrevalenceFields' isEditable attribute. I can submit this change in a separate PR as it's not related to this PR.

Oh good catch. Yes, happy to have a PR on that issue as well. 👍

blanchardjeremy commented:
And one last fix. 

![image](https://user-images.githubusercontent.com/32290/104678795-1a8b6b00-56a1-11eb-96b2-df98d1cd0477.png)


After the user types in the manual data, I think they should see the actual prevalence and adjusted prevalence numbers still. The live site has the functionality.

david-mears-dfe commented:
> And one last fix.
> 
> After the user types in the manual data, I think they should see the actual prevalence and adjusted prevalence numbers still. The live site has the functionality.

Are you sure? I can't see this behaviour on microcovid.org. But it is simple to add. The change would be from:
```
{!isManualEntryCurrently && <PrevalenceResult data={data} />}
```
to:
```
<PrevalenceResult data={data} />
```

@blanchardjeremy 

<img width="427" alt="Screenshot 2021-01-15 at 18 35 01" src="https://user-images.githubusercontent.com/60350599/104765174-63562900-5760-11eb-8497-eac91682c69f.png">

blanchardjeremy commented on src/locales/en.json:
```diff
@@ -109,9 +109,12 @@
     "location_subprompt_US-AK": "Entire state, or select borough...",
     "location_subprompt_US": "Entire state, or select county...",
-    "location_selector_header": "Step 1: Enter your location",
+    "location_selector_header": "Step 1: Location Risk",
```
For this one, I'd prefer to opt for "Enter your location" so it is a clear instruction. 

I can see why you made this change. If you'd prefer, I think it's fine for this header to change when switched to manual data mode.

Whatcha think?

blanchardjeremy commented:
> Are you sure? I can't see this behaviour on microcovid.org. But it is simple to add. The change would be from:

Oh you're right. It's not that way on the live site.... Huh. I swore I got that behavior last night! *shrug*  Yeah, it's not a blocking issue, but my two cents is that we should include it. 

So yeah. I think we're ready to go after my comment about the "Step 1" header above.

Thanks for all you've done! 🙌

blanchardjeremy commented:
@david-mears-dfe I'd love to merge this into main before merging #646 so that you don't have extra work to do. If you have a chance, could you make the change to "step 1" and sync it with the base branch? Then I can merge it into main!

beshaya commented:
Landed as #656

beshaya commented:
Thanks for the PR, @david-mears-dfe! I'm really sorry that took so long.

david-mears-dfe commented:
Nice, thanks both.

blanchardjeremy commented:
Yeah. Thanks for both writing the code and for the feedback. 👍✅
